### PR TITLE
feat: Add PyTorch Word language Modeling example to Determined's Example [DET-5226]

### PR DESCRIPTION
**Pull request by RehanSD**
## Description
Add the PyTorch Word Language Modeling example, found [here](https://github.com/pytorch/examples/tree/master/word_language_model) to Determined's examples.

Resolves [DET-5226]

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234
[DET-5226]: https://determinedai.atlassian.net/browse/DET-5226

### CODE REVIEW

File: examples/nlp/word_language_model/model_def.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        hparams = self.context.get_hparams()
+        self.using_bind_mount = data_config.get("use_bind_mount", False)
+        self.use_cache = data_config.get("use_cache", True)
+        self.eval_batch_size = data_config.get("eval_batch_size", 10)
+        self.bind_mount_path = (
+            Path(data_config.get("bind_mount_path")) if self.using_bind_mount else None
+        )
+
+        download_directory = (
+            self.bind_mount_path if self.using_bind_mount else Path("/tmp")
```
Comment (naren-determined) [inline examples/nlp/word_language_model/model_def.py:40]: Why do you need this `Path("/tmp")` here. Can't you directly set else as the `f"data-rank{self.context.distributed.get_rank()}`?

**Review by naren-determined — CHANGES_REQUESTED** [review]

File: e2e_tests/tests/nightly/test_capability.py [hunk 2]
```diff
@@ -52,3 +52,43 @@ def test_deformabledetr_coco_pytorch_const() -> None:
     exp.run_basic_test_with_temp_config(
         config, conf.cv_examples_path("deformabledetr_coco_pytorch"), 1
     )
+
+
+@pytest.mark.nightly  # type: ignore
```
[the remark this answers is not in the document]
Comment (RehanSD) [reply e2e_tests/tests/nightly/test_capability.py:57]: I think we could do just testing the Transformer and the RNN? The RNN, LSTM, and GRU cells all share a codepath, so if RNN is working, the others should as well? The only reason one would work but the other two don't is if there is a breaking change to the underlying PyTorch implementations for the cell.

File: examples/nlp/word_language_model/model_def.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+class WordLanguageModelPyTorch(PyTorchTrial):
+    def __init__(self, context: PyTorchTrialContext):
+        self.context = context
+        data_config = self.context.get_data_config()
+        hparams = self.context.get_hparams()
+        self.using_bind_mount = data_config.get("use_bind_mount", False)
```
[the remark this answers is not in the document]
Comment (RehanSD) [reply examples/nlp/word_language_model/model_def.py:32]: For this, my reasoning was that I didn't want users to have to specify all of the numbers, especially since many of them have defaults (the original values in the pytorch repo). I can go ahead and specify them in the .yaml files if you'd prefer, instead!

File: examples/nlp/word_language_model/model_def.py (hunk 1 above)
Comment (RehanSD) [reply examples/nlp/word_language_model/model_def.py:40]: Does this mean that the default location for the data should not be `/tmp` but rather `/data`?

File: e2e_tests/tests/nightly/test_capability.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (naren-determined) [reply e2e_tests/tests/nightly/test_capability.py:57]: Yes let's test just Transformer and RNN then please.

File: examples/nlp/word_language_model/const.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    smaller_is_better: true
+min_validation_period:
+    epochs: 1
+data:
+    use_bind_mount: True
+    bind_mount_path: /mnt/data
+entrypoint: model_def:WordLanguageModelPyTorch
+bind_mounts:
+    - host_path: /tmp/
+      container_path: /mnt/data
```
[the remark this answers is not in the document]
Comment (naren-determined) [reply examples/nlp/word_language_model/const.yaml:26]: Oh sorry meant setting `container_path` to `/data`, host_path can still be `/tmp` or `/data` whatever you prefer.

File: examples/nlp/word_language_model/model_def.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (naren-determined) [reply examples/nlp/word_language_model/model_def.py:32]: You can just set the numbers in the .yaml files though right? Either way, the users would have to specify them. I think it makes for cleaner code to have all the config settings reside in the .yaml so let's do that.

File: examples/nlp/word_language_model/model_def.py (hunk 1 above)
Comment (naren-determined) [reply examples/nlp/word_language_model/model_def.py:40]: Yeah I mean this is whatever you prefer. I just want some level of parity between the dtrain path and the single GPU path. Whether it's /data or /tmp, I just want that to be consistent.

File: examples/nlp/word_language_model/const.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (RehanSD) [reply examples/nlp/word_language_model/const.yaml:26]: Using /data causes agents to crash - if its ok with you, I'll use /mnt/data?

**Review by naren-determined — COMMENTED** [review]
Can you remove the RNN testing/related code since we decided we are not going to showcase that example? For the distributed training cases -- if you are getting good performance I wouldn't worry about scaling the learning rate but if you aren't try that out as well.

**Comment by RehanSD** [comment]
> Can you remove the RNN testing/related code since we decided we are not going to showcase that example? For the distributed training cases -- if you are getting good performance I wouldn't worry about scaling the learning rate but if you aren't try that out as well.

The RNN testing code is for the LSTM/GRU models - should I still remove it?

**Review by naren-determined — APPROVED** [review]
Lgtm 🚢
